### DISCUSSION TITLE
Update Collaborator Permission

### DISCUSSION THREAD

**Discussion by chrissphinx** [community/community#24738] [Apps, API and Webhooks]
<p>Is there any way to update a collaborators permission on an organization repository via the v3 (or even v4) API?</p>
<p>I cannot find any way to do this and it would very much help me enforce assignment deadlines in GitHub Classroom, since Classroom does not do anything in terms of deadline enforcement. I can make the switch for student collaborators from <strong>Write</strong> to <strong>Read</strong> permissions just fine on the website but would be quite cumbersome to do that for 70+ repositories.</p>

**Comment by chrissphinx** [comment, marked as answer]
<p>Okay, so this turns out to be an API documentation issue because: <a href="https://developer.github.com/v3/repos/collaborators/#add-user-as-a-collaborator" rel="nofollow noopener">https://developer.github.com/v3/repos/collaborators/#add-user-as-a-collaborator</a></p>
<p>Nowhere in the docs here is it mentioned that this endpoint can also update the permissions on existing collaborators, although it is mentioned in the Octokit library docs: <a href="https://github.com/octokit/octokit.rb/blob/14bb1a1d4fec8d2250dc6f6b9cb2d446a0332d58/lib/octokit/client/repositories.rb#L337" rel="nofollow noopener">https://github.com/octokit/octokit.rb/blob/14bb1a1d4fec8d2250dc6f6b9cb2d446a0332d58/lib/octokit/client/repositories.rb#L337</a></p>
<p>But, even more confusingly, the given permission levels do not match the levels shown on the GitHub website. There are actually <strong>five</strong> permission levels, not three: Read, Triage, Write, Maintain, Admin.</p>
<p><strong>Even worse</strong> , when hitting the endpoint using the permissions “Triage”, “Maintain” or “Admin”, they successfully modify the existing collaborator permissions. But, when using what you would might expect to work (“Read” or “Write”) it will <em>fail</em>. You need to use the permissions “Pull” (which is read-only access or “Read” permissions on the website) or “Push” (which is “Write” permissions on the website) for this to work properly.</p>
<p>So either the permission names on the GitHub website need to be changed, or the documentation for the API needs to updated to explain that:</p>
<ol>
<li>the add user as a collaborator endpoint can actually update the permission of existing users and</li>
<li>the actual permission levels allowed are: Pull (which is Read), Triage, Push (which is Write), Maintain, Admin</li>
</ol>
<p>Though, it would be really awesome to allow “Read” and “Write” on this endpoint as alises for the currently established names (according to the API anyways) to maybe save someone down the line a huge, hours-long headache.</p>

**Comment by jandiorio** [comment]
<p>This was very useful.  Working with the API to add collaborators I have found that setting the initial permission level doesn’t work. Have you seen and/or solved that issue.  </p>
<p>permission=pull doesn’t set the initial level to Read in my experience. </p>
<p>Thanks</p>

**Comment by bennn** [comment]
<p>I haven’t been able to assign “pull” access. Not initially, and not with later calls.</p>
<p>Push access works fine. Haven’t tried any others.</p>

**Comment by niklas-ourmachinery** [comment]
<p><a class="mention" href="/jandiorio">@jandiorio</a> <a class="mention" href="/bennn">@bennn</a></p>
<p>It’s a bit confusing. If you look at the documentation it says:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/b/b/bb47cb67ee4056889683a88df2c9864f556d1655.png" data-download-href="/uploads/short-url/qIL6zQ4Tvkgp5Be5b0fqMT9yGJD.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/26548514/181086510-06039c32-6465-4250-8853-9b5bc83fdaff.png" alt="image" data-base62-sha1="qIL6zQ4Tvkgp5Be5b0fqMT9yGJD" width="690" height="485" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/b/b/bb47cb67ee4056889683a88df2c9864f556d1655_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">771×543 24.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>See the <code>body</code> label for <code>permission</code> and <code>permissions</code>. What this means that these parameters need to be passed as JSON in the BODY of the HTTP request, not as query parameters in the URL. I missed this at first too and scratched my head for a while trying to figure out why it was not working.</p>
<p>So instead of specifying <code>?permission=pull</code> in the URL you should use the following BODY in the HTTP request:</p>
<pre><code class="lang-auto">{"permission": "pull"}
</code></pre>
<p>If you do that, the newly added collaborator will get the <strong>Read</strong> role.</p>

**Reply by bennn** [reply]
Thanks. I did use JSON back then. Hope it's been fixed.

**Comment by rutgerc-klarrio** [comment]
It is very confusing.  Different semantics are used for the website UI but the API is also inconsistent:

When POST'ing a new team permission with permission 'maintain' to `https://api.github.com/orgs/ORG/teams/TEAM/repos/OWNER/REPO` following the documentation [here](https://docs.github.com/en/rest/teams/teams#add-or-update-team-repository-permissions) and afterwards GET'ing the teams via `https://api.github.com/repos/OWNER/REPO/teams` the `permission` field for the team will be `push` and not `maintain`.  The individual levels in the `permissions` field will be correct (`"maintain": true`).
